WalletWasabi/Tor/Socks5/TorSocks5ClientFactory.cs

```diff
@@ -0,0 +1,408 @@
+				tcpClient = new TcpClient(TorSocks5EndPoint.AddressFamily);
+
+				transportStream = await ConnectAsync(tcpClient, cancellationToken).ConfigureAwait(false);
+				await HandshakeAsync(tcpClient, isolateStream, cancellationToken).ConfigureAwait(false);
+				await ConnectToDestinationAsync(tcpClient, host, port, cancellationToken).ConfigureAwait(false);
+
+				if (useSsl)
+				{
+					transportStream = await UpgradeToSslAsync(tcpClient, host).ConfigureAwait(false);
+				}
+
+				var result = new TorConnection(tcpClient, transportStream);
+
+				transportStream = null;
+				tcpClient = null;
+				return result;
+			}
+			catch (TorException)
+			{
+				throw;
+			}
+			finally
+			{
+				transportStream?.Dispose();
+				tcpClient?.Dispose();
+			}
+		}
+
+		private async static Task<SslStream> UpgradeToSslAsync(TcpClient tcpClient, string host)
+		{
+			var sslStream = new SslStream(tcpClient.GetStream(), leaveInnerStreamOpen: false);
+			await sslStream.AuthenticateAsClientAsync(host, new X509CertificateCollection(), IHttpClient.SupportedSslProtocols, true).ConfigureAwait(false);
+			return sslStream;
+		}
+
+		/// <summary>
+		/// Establishes TCP connection with Tor SOCKS5 endpoint.
+		/// </summary>
+		/// <exception cref="ArgumentException">This should never happen.</exception>
+		/// <exception cref="TorException">When connection to Tor SOCKS5 endpoint fails.</exception>
+		private async Task<NetworkStream> ConnectAsync(TcpClient tcpClient, CancellationToken cancellationToken = default)
+		{
+			if (!TorSocks5EndPoint.TryGetHostAndPort(out string? host, out int? port))
+			{
+				throw new ArgumentException("Endpoint type is not supported.", nameof(TorSocks5EndPoint));
+			}
+
+			try
+			{
+				await tcpClient.ConnectAsync(host, port.Value, cancellationToken).ConfigureAwait(false);
+				return tcpClient.GetStream();
+			}
+			catch (Exception ex) when (IsConnectionRefused(ex))
+			{
+				throw new TorConnectionException($"Could not connect to Tor SOCKSPort at '{host}:{port}'. Is Tor running?", ex);
+			}
+		}
+
+		/// <summary>
+		/// Do the authentication part of Tor's SOCKS5 protocol.
+		/// </summary>
+		/// <param name="isolateStream">Whether random username/password should be used for authentication and thus effectively create a new Tor circuit.</param>
+		/// <remarks>Tor process must be started with enabled <c>IsolateSOCKSAuth</c> option. It's ON by default.</remarks>
+		/// <seealso href="https://www.torproject.org/docs/tor-manual.html.en"/>
+		/// <seealso href="https://linux.die.net/man/1/tor">For <c>IsolateSOCKSAuth</c> option explanation.</seealso>
+		/// <seealso href="https://gitweb.torproject.org/torspec.git/tree/socks-extensions.txt#n35"/>
+		/// <exception cref="TorAuthenticationException">When authentication fails due to unsupported authentication method or invalid credentials.</exception>
+		private async Task HandshakeAsync(TcpClient tcpClient, bool isolateStream = false, CancellationToken cancellationToken = default)
```

Comment:
`TcpClient` is used because `tcpClient.ReceiveBufferSize` is used in `SendAndReceiveAsync`. So I would need to pass two parameters instead of one. Or I would need to refactor it more.

Reply:
I understand. I don't think touching it more is a good idea.